### PR TITLE
Fix initContainer overrides

### PR DESCRIPTION
**Pull request by fanny-jiang**
### What does this PR do?

Fix container overrides for init containers: 

* name
* resources
* volumeMounts
* env vars
* securityContext (not including secomp profile)

Some container overrides were omitted because they are not supported in init containers (livenessProbe and readinessProbe), may cause regression in functionality (args, command), or would have no effect (DD healthPort). 

This PR also fixes override container unit tests as well as adds new unit tests for init container overrides. 

### Motivation

Bug fix and customer support case

### Additional Notes

Anything else we should know when reviewing?

### Minimum Agent Versions

Are there minimum versions of the Datadog Agent and/or Cluster Agent required?

* Agent: N/A
* Cluster Agent: N/A

### Describe your test plan
Test in kind cluster and non-kind cluster:
1. Deploy operator
2. Apply `DatadogAgent` manifest overriding init container configs
3. Check that overrides are applied: `kubectl get pod -oyaml <agent_pod_name>`
 
```
apiVersion: datadoghq.com/v2alpha1
kind: DatadogAgent
metadata:
  name: datadog
spec:
  global:
    clusterName: kind-kind  # needed for kind
    kubelet:
      tlsVerify: false  # needed for kind
    credentials:
      apiSecret:
        secretName: datadog-secret
        keyName: api-key
      appSecret:
        secretName: datadog-secret
        keyName: app-key
#  features:  # can't test system-probe in kind
#    npm:
#      enabled: true
  override:
    nodeAgent:
      containers:
        agent:
          env:
            - name: DD_LOG_LEVEL
              value: debug
        init-config:
          name: new-init-config-name
          env:
            - name: INIT_ENV_VAR
              value: SOME_VALUE
          resources:
            limits:
              cpu: "3"
              memory: "2Gi"
            requests:
              cpu: "2"
              memory: "1Gi"
        init-volume:
          resources:
            limits:
              cpu: "3"
              memory: "2Gi"
            requests:
              cpu: "2"
              memory: "1Gi"
          volumeMounts:
          - name: custom-volume
            mountPath: /custom-volume
          securityContext:
            runAsUser: 0
#        seccomp-setup:  # can't test system-probe in kind
#          securityContext:
#            runAsUser: 0
      volumes:
        - emptyDir: {}
          name: custom-volume
```

### Checklist

- [x] PR has at least one valid label: `bug`, `enhancement`, `refactoring`, `documentation`, `tooling`, and/or `dependencies`
- [x] PR has a milestone or the `qa/skip-qa` label

### CODE REVIEW

File: controllers/datadogagent/merger/fake/envvar_manager.go [hunk 1]
```diff
@@ -24,14 +24,27 @@ func (_m *EnvVarManager) AddEnvVar(newEnvVar *v1.EnvVar) {
...
-func (_m *EnvVarManager) AddEnvVarToInitContainer(initContainerName commonv1.AgentContainerName, newEnvVar *v1.EnvVar) {
-	_m.t.Logf("AddEnvVar %s: %#v", newEnvVar.Name, newEnvVar.Value)
-	_m.EnvVarsByC[initContainerName] = append(_m.EnvVarsByC[initContainerName], newEnvVar)
+func (_m *EnvVarManager) AddEnvVarToInitContainer(containerName commonv1.AgentContainerName, newEnvVar *v1.EnvVar) {
+	for _, initContainerName := range initContainerNames {
+		if containerName == initContainerName {
+			_m.t.Logf("AddEnvVar to container %s key:%s value:%#v", containerName, newEnvVar.Name, newEnvVar.Value)
+			_m.EnvVarsByC[containerName] = append(_m.EnvVarsByC[containerName], newEnvVar)
+		}
+	}
```
Comment (fanny-jiang) [inline controllers/datadogagent/merger/fake/envvar_manager.go:47]: These changes are intended to mimic the behavior of the [actual envvar manager](https://github.com/DataDog/datadog-operator/blob/480f69931540a7de1808ffd7bd9b5043c1249fe7/controllers/datadogagent/merger/envvars.go#L56-L86), which checks the container names before appending the overrides to the found container's env vars list

File: controllers/datadogagent/merger/fake/volume_mount_manager.go [hunk 2]
```diff
@@ -23,12 +25,25 @@ func (_m *VolumeMountManager) AddVolumeMount(volumeMount *v1.VolumeMount) {
...
 }
 
 // AddVolumeMountToInitContainer provides a mock function with given fields: volumeMount, containerName
 func (_m *VolumeMountManager) AddVolumeMountToInitContainer(volumeMount *v1.VolumeMount, containerName commonv1.AgentContainerName) {
-	_m.VolumeMountsByC[containerName] = append(_m.VolumeMountsByC[containerName], volumeMount)
+	for _, initContainerName := range initContainerNames {
+		if containerName == initContainerName {
+			_m.VolumeMountsByC[containerName] = append(_m.VolumeMountsByC[containerName], volumeMount)
+		}
+	}
```
Comment (fanny-jiang) [inline controllers/datadogagent/merger/fake/volume_mount_manager.go:46]: Same as the above mocked envvar manager https://github.com/DataDog/datadog-operator/pull/893/files#r1302201156, but with the mocked volumeMount manager. Actual volumeMount manager [behavior](https://github.com/DataDog/datadog-operator/blob/480f69931540a7de1808ffd7bd9b5043c1249fe7/controllers/datadogagent/merger/volume_mount.go#L43-L57).

File: controllers/datadogagent/override/container_test.go [hunk 3]
```diff
@@ -7,54 +7,79 @@ package override
...
 				for _, container := range manager.PodTemplateSpec().Spec.Containers {
-					if container.Name == string(commonv1.CoreAgentContainerName) {
+					if container.Name == "my-container-name" {
+						foundMatching = true
 						assert.Equal(t, "my-container-name", container.Name)
+						break
 					}
 				}
+				if !foundMatching {
+					t.Errorf("Could not find matching container: %s", containerName)
```
Comment (levan-m) [inline controllers/datadogagent/override/container_test.go:65]: This isn't really an error right condition right, rather a failed unit test? If yes, probably better to use `assert.True(..)`

File: controllers/datadogagent/override/container_test.go [hunk 4]
```diff
@@ -62,7 +62,7 @@ func TestContainer(t *testing.T) {
 					}
 				}
 				if !foundMatching {
```
Comment (levan-m) [inline controllers/datadogagent/override/container_test.go:64]: nit - if check is redundant

**Review by levan-m — APPROVED** [review]

**Pull request merged**